packages/controllers/src/subject-metadata/SubjectMetadataController.ts

```diff
@@ -0,0 +1,261 @@
+import {
+  BaseControllerV2 as BaseController,
+  RestrictedControllerMessenger,
+} from '@metamask/controllers';
+import type { Patch } from 'immer';
+import { Json } from 'json-rpc-engine';
+
+import type {
+  GenericPermissionController,
+  PermissionSubjectMetadata,
+  HasPermissions,
+} from '../permissions';
+
+const controllerName = 'SubjectMetadataController';
+
+type SubjectOrigin = string;
+
+export type SubjectMetadata = PermissionSubjectMetadata & {
+  [key: string]: Json;
+  name: string;
+  // TODO:TS4.4 make optional
+  extensionId: string | null;
+  host: string | null;
+  iconUrl: string | null;
+};
+
+type SubjectMetadataToAdd = PermissionSubjectMetadata & {
+  name: string;
+  extensionId?: string | null;
+  host?: string | null;
+  iconUrl?: string | null;
+} & Record<string, Json>;
+
+export type SubjectMetadataControllerState = {
+  subjectMetadata: Record<SubjectOrigin, SubjectMetadata>;
+};
+
+const stateMetadata = {
+  subjectMetadata: { persist: true, anonymous: true },
+};
+
+const defaultState: SubjectMetadataControllerState = {
+  subjectMetadata: {},
+};
+
+export type GetSubjectMetadataState = {
+  type: `${typeof controllerName}:getState`;
+  handler: () => SubjectMetadataControllerState;
+};
+
+export type ClearSubjectMetadataState = {
+  type: `${typeof controllerName}:clearState`;
+  handler: () => void;
+};
+
+export type TrimSubjectMetadataState = {
+  type: `${typeof controllerName}:trimMetadataState`;
+  handler: () => void;
+};
+
+export type SubjectMetadataControllerActions =
+  | GetSubjectMetadataState
+  | ClearSubjectMetadataState
+  | TrimSubjectMetadataState;
+
+export type SubjectMetadataStateChange = {
+  type: `${typeof controllerName}:stateChange`;
+  payload: [SubjectMetadataControllerState, Patch[]];
+};
+
+export type SubjectMetadataControllerEvents = SubjectMetadataStateChange;
+
+type AllowedActions = HasPermissions;
+
+export type SubjectMetadataControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  SubjectMetadataControllerActions | AllowedActions,
+  SubjectMetadataControllerEvents,
+  AllowedActions['type'],
+  never
+>;
+
+type SubjectMetadataControllerOptions = {
+  messenger: SubjectMetadataControllerMessenger;
+  subjectCacheLimit: number;
+  state?: Partial<SubjectMetadataControllerState>;
+};
+
+/**
+ * A controller for storing metadata associated with permission subjects. More
+ * or less, a cache.
+ */
+export class SubjectMetadataController extends BaseController<
+  typeof controllerName,
+  SubjectMetadataControllerState,
+  SubjectMetadataControllerMessenger
+> {
+  private _subjectCacheLimit: number;
+
+  get subjectCacheLimit(): number {
```

Comment:
What is this for?

Reply:
Hrm, it's not necessary. Removed in: f7bd1cc